Use `new (nothrow)` everywhere

peter-b opened the pull request:
We jump through various hoops in order to get new to be recognised as
having nothrow behaviour but it is fragile and doesn't work when the
engine is hosted inside another process (if that process uses its own
implementation of new, as it is entitled to do).

Instead of dealing with this, just bite the bullet and add the
(std::nothrow) argument to every invocation of new in the engine and
externals.

Updated/rebased version of #4593

livecodefraser committed:
Remove the now-redundant '-fcheck-new' compile option

We use `new (nothrow)` now so it isn't needed.

livecodefraser committed:
Add `using std::nothrow;` to core.h

All uses of new now consistently use (nothrow) as the placement
parameter instead of mixing it with (std::nothrow).

livecodefraser committed:
Use `new (nothrow)` everywhere

We jump through various hoops in order to get new to be recognised as
having nothrow behaviour but it is fragile and doesn't work when the
engine is hosted inside another process (if that process uses its own
implementation of new, as it is entitled to do).

Instead of dealing with this, just bite the bullet and add the
(std::nothrow) argument to every invocation of new in the engine and
externals.

peter-b committed:
libexternal: Add `using std::nothrow` to Unix support lib

peter-b committed:
Add requirement for `new` with placement or `nothrow` to C++ style

peter-b commented:
This needs an update to `CONTRIBUTING.md` to mention that all uses of `new` need to either be `(nothrow)` or specify placement.

livecodeali commented on engine/src/object.h:
```diff
@@ -450,7 +450,8 @@ struct MCPatternInfo
-class MCObject : public MCDLlist
+class MCObject :
+  public MCDLlist
 {
```
Some extraneous whitespace changes here, probably not worth reverting but it proves I was looking properly, right ;-)

livecodeali commented:
@livecode-vulcan review ok 85a7444

livecode-vulcan commented:
💙 review by @livecodeali ok 85a744465b3a4e05d90228bc38efb8aa574322eb

livecode-vulcan commented:
😞 test failure 85a744465b3a4e05d90228bc38efb8aa574322eb
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/326)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/331)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/326)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/326)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/332)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/326)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/326)

livecodefraser committed:
Remove lots of unneeded definitions from core.h

All out platforms supply the required C standard library headers;
just use them directly instead.

livecodefraser committed:
Fix the debugging 'new' defines for Windows

They didn't supply the correct overloads, breaking most uses of 'new'
when compiling in debug mode.

livecodefraser committed:
Ensure files using `new (nothrow)` include appropriate headers

livecodefraser committed:
Fix some uses of 'nil' as an integer constant

peter-b committed:
The null X11 resource is None, not a null pointer

`X11/X.h` defines both `XID` (of which the X11 `Window` type is an
alias), and the `None` constant which represents the null resource or
atom.

peter-b committed:
Merge branch 'develop' into nothrow_new

The following conflicts, due to Coverity defect fixes by
@montegoulding that touched some of the same lines, were resolved by
taking the `develop` version and adding `(nothrow)` as applicable:

	engine/src/hc.cpp
	engine/src/opensslsocket.cpp

livecodeali commented:
@livecode-vulcan review ok 97b61d1

livecode-vulcan commented:
💙 review by @livecodeali ok 97b61d1c0aaa814d9f6d4795992cac7abd0d0734

livecode-vulcan commented:
😎 test success 97b61d1c0aaa814d9f6d4795992cac7abd0d0734
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/413)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/418)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/413)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/413)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/419)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/413)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/413)
